File: packages/cli-python/src/spec_init/persona.py
"""Strip persona-gated blocks; byte-parity Python port of strip-personas.mjs."""

from __future__ import annotations

import re

from .args import Persona

_VALID: frozenset[str] = frozenset({"vibe", "student", "engineer", "team"})

_BLOCK_RE = re.compile(
    r"<!--\s*persona:([a-z,\s]+)\s*-->([\s\S]*?)<!--\s*/persona:\1\s*-->\n?"
)


def strip_personas(source: str, persona: Persona) -> str:
    """Strip persona-gated blocks from `source`, keeping only those matching `persona`."""
    if persona not in _VALID:
        raise ValueError(f"unknown persona: {persona}")

    def _replace(match: re.Match[str]) -> str:
        names_raw = match.group(1)
        body = match.group(2)
        names = [n.strip() for n in names_raw.split(",") if n.strip()]
        for name in names:
            if name not in _VALID:
                raise ValueError(f"unknown persona in marker: {names_raw.strip()}")
        return body if persona in names else ""

    stripped = _BLOCK_RE.sub(_replace, source)
    # Collapse 3+ newlines left by removed blocks; end with one trailing newline.
    stripped = re.sub(r"\n{3,}", "\n\n", stripped)
    stripped = re.sub(r"\n+$", "\n", stripped)
    return stripped

File: packages/cli-node/src/persona.ts
// Strip persona-gated blocks; byte-parity TS port of tools/strip-personas.mjs.

import type { Persona } from './args.js';

const VALID: ReadonlySet<string> = new Set(['vibe', 'student', 'engineer', 'team']);

/** Strip persona-gated blocks from `source`, keeping only those matching `persona`. */
export function stripPersonas(source: string, persona: Persona): string {
  if (!VALID.has(persona)) {
    throw new Error(`unknown persona: ${persona}`);
  }
  const blockRe = /<!--\s*persona:([a-z,\s]+)\s*-->([\s\S]*?)<!--\s*\/persona:\1\s*-->\n?/g;
  const stripped = source.replace(blockRe, (_match, namesRaw: string, body: string) => {
    const names = namesRaw
      .split(',')
      .map((n) => n.trim())
      .filter(Boolean);
    if (names.some((n) => !VALID.has(n))) {
      throw new Error(`unknown persona in marker: ${namesRaw.trim()}`);
    }
    return names.includes(persona) ? body : '';
  });
  // Collapse 3+ newlines left by removed blocks; end with one trailing newline.
  return stripped.replace(/\n{3,}/g, '\n\n').replace(/\n+$/, '\n');
}

const ACTIVE_PERSONA_RE = /^\/\/ active-persona:.*$/m;
const MANAGED_SUFFIX = '   // managed by spec-init - change via `spec-init customize --persona`';

/** Replace the CLI-managed `// active-persona:` line with `persona` (or `<none>`). No-op if absent. */
export function stampPersona(text: string, persona: Persona | null): string {
  const line = `// active-persona: ${persona ?? '<none>'}${MANAGED_SUFFIX}`;
  return ACTIVE_PERSONA_RE.test(text) ? text.replace(ACTIVE_PERSONA_RE, line) : text;
}
